### DISCUSSION TITLE
Ability to lock palettes to a particular schema

### DISCUSSION THREAD

**Discussion by stevenpetryk** [primer/prism#41] [Ideas]
I think it'd be really neat if any palette could be forced to fit a particular shape. Most design systems don't willy-nilly add and remove scales, and most scales have a set number of colors.

For a real world example: at Discord we're adopting Prism into our workflow in such a way that designers will be able to make live tweaks to our underlying theme colors and see them reflected in the app. We have Discord embedded in an iframe and are using `postMessage` to send palette updates at runtime:

![](https://user-images.githubusercontent.com/1724000/175376311-7621b8b8-73e6-4094-ba95-9caa4d285e2f.png)

For situations like ours, where we're planning to use Prism as a direct input into our code (and thus into our color pipeline), we'd really _never_ want folks to add or delete scales, or to change the number of colors in a scale.

So my thinking is that there could be the concept of a "schema", which lets one say that a palette should have a certain set of scales and a certain number of colors for each of those scales. It would be impossible to add or remove scales/colors when a schema is applied to a palette.

@colebemis we plan on adding this to our fork, but if it's something you think would be useful in general, we'd be happy to open a PR!

---

Side-note, are issues like this useful? We're going to be adding a lot of Discord-specific features that may be generalizable to the tool. (For example, you can see our "apply easing function" functionality in that screenshot).

**Comment by stevenpetryk** [comment]
As a followup, it might be nice if the schema came with a default state that the palette could be reset back to.

**Comment by colebemis** [comment]
> at Discord we're adopting Prism into our workflow in such a way that designers will be able to make live tweaks to our underlying theme colors and see them reflected in the app. We have Discord embedded in an iframe and are using postMessage to send palette updates at runtime

🤯 Mind blown. This looks amazing. I'd love to see a video demo of how this works.

> So my thinking is that there could be the concept of a "schema", which lets one say that a palette should have a certain set of scales and a certain number of colors for each of those scales. It would be impossible to add or remove scales/colors when a schema is applied to a palette.

I like the concept of a palette "schema." I'd be curious to see how you imagine it would work in the UI. I'd also want to ensure that it's opt-in. In other words, if someone creates a new palette, they don't need to define a schema unless they want to.

> Side-note, are issues like this useful? We're going to be adding a lot of Discord-specific features that may be generalizable to the tool. (For example, you can see our "apply easing function" functionality in that screenshot).

Yes! I love seeing how folks are using this tool. I think the best ideas will come from using the tool in practice.

~~However, I'm currently the only one maintaining Prism and I'd like to avoid becoming a bottleneck for your team. I'd recommend working from a fork (if you aren't already) so y'all can move as fast as you need.~~ Update: I just re-read your original comment and it looks like you're already working from a fork. Perfect 👍
